Build command does incremental compilation

cdmihai opened the pull request:
Introduce incremental compilation for Build that passes the basic E2E helloworld tests:

- Build and Compile share the code that gets them input or output paths
- Build uses the shared code to check if there are any inputs that are newer than the oldest output
- Added some end to end tests that assert a few incremental scenarios
- Introduced a new --force-incremental-unsafe flag for Build, if users wish to mark a build as unsafe (and thus prevent incremental compilation)

CLI itself still does not self host with incrementality because the postcompile `place-binary` script prevents it. A future PR will address CLI incremental selfhost.

dnfclas commented:
Hi __@cdmihai__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

davidfowl commented on scripts/test/runtests.ps1:
```diff
@@ -44,7 +49,7 @@ pushd "$TestBinRoot"
 # Run each test project
 $TestProjects | ForEach-Object {
-    & "corerun.exe"  "xunit.console.netcore.exe" "$_.dll" -xml "$_.xml" -notrait category=failing
+    & "corerun.exe"  "xunit.console.netcore.exe" "$_.dll" -xml "$_.xml" -notrait category=failing -parallel none
```
Why are you turning off parallel tests?

cdmihai replied:
I was having some issues with tests so I preventively turned it off. However it turned out my tests were failing because the stage 0 was used and not stage 2 :)

I left them turned off as a safety measure in case tests from different test classes might read / write on the same paths which would lead to races and weird test failures.
Until CLI tests use a shared mechanism that guarantees the absence file IO races, I think it's safer to not have parallel tests.

davidfowl replied:
If our tests are written poorly I agree it'll happen. I'd revert this change as its unrelated to your changes.

davidfowl replied:
He stage 0 stage 2 thing is a bug and should be fixed

piotrpMSFT replied:
Agreed. Let's leave parallel on. If someone writes a buggy test they will fix it.

cdmihai replied:
Alright. I'll leave them turned off.

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -23,25 +26,25 @@ internal class CompileContext
             _rootProject = rootProject;
-            _args = args;
+
+            //Cleaner to clone the args and mutate the clone than have separate CompileContext fields for mutated args 
```
nit: Put a space between `// and the comment`

davidfowl commented on src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs:
```diff
@@ -113,6 +115,46 @@ public static string GetResourceFileMetadataName(Project project, KeyValuePair<s
+        //used in incremental compilation
+        public static IEnumerable<string> GetCompilationSources(ProjectContext project) => project.ProjectFile.Files.SourceFiles;
+
+
```
What's with all the extra space?

cdmihai replied:
Probably a remnant from moving a lot of code around. Thanks for pointing that out

davidfowl commented on src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs:
```diff
@@ -113,6 +115,46 @@ public static string GetResourceFileMetadataName(Project project, KeyValuePair<s
+
+
+        //used in incremental compilation
+        public static string GetCompilationOutput(Project project, NuGetFramework framework, string configuration, string outputPath)
```
Doesn't this helper exist already? /cc @piotrpMSFT

cdmihai replied:
I moved it here from Compile.

davidfowl commented on test/E2E/E2ETest.cs:
```diff
@@ -50,6 +51,76 @@ public void TestDotnetCompile()
             TestOutputExecutable(OutputDirectory);
         }
 
+        [Fact] public void TestDotnetIncrementalBuild()
```
Fix the formatting. Put Fact on top of the method like the rest of the tests.

cdmihai replied:
Good catch. Thanks.

davidfowl commented on src/Microsoft.DotNet.Tools.Compiler/CompilerUtil.cs:
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
This class is beginning to look like a dumping ground for helper methods...

cdmihai replied:
I am using this class to hold the functionality that needs to be available to both build and compile. Currently I'm using it to learn and collect what the common interface is. Ideally, this communication should be done via something more first class than a utility class (an interface maybe?).

For now, I'd rather keep it like this until we get a better idea on how the 2 verbs need to communicate.

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/ProjectDependenciesFacade.cs:
```diff
@@ -0,0 +1,56 @@
+        public Dictionary<string, ProjectDescription> ProjectDependenciesWithSources { get; }
+        public List<LibraryExport> Dependencies { get; }
+
+        public ProjectDependenciesFacade(ProjectContext _rootProject, string configValue)
```
Why is there an _ in the parameter name?

cdmihai replied:
I rewrote the way that class got instantiated a few times. A field name must have slipped in. Thanks for pointing that out.

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
+        {
+            var compilerIO = GetCompileIO(project, _args.ConfigValue, _args.OutputValue, _args.IntermediateValue, dependencies);
+
+            if (! (compilerIO.Outputs.Any() && compilerIO.Inputs.Any()))
```
Couldn't you do all of this logic with a single for loop over inputs and outputs?

cdmihai replied:
True. However, right now I am still fleshing out the desired behaviour on edge cases and it helps to have them separated. I'd leave this for future refactorings when specs are more clear.

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
-                if (!InvokeCompileOnDependency(dependency, _outputPath, _intermediateOutputPath))
+                if (incremental)
+                {
+                    var dependencyContext = ProjectContext.Create(dependency.Path, dependency.TargetFrameworkInfo.FrameworkName);
```
`dependency.TargetFrameworkInfo.FrameworkName` isn't correct. It should be using `dependency.Framework`

cdmihai replied:
OK. I wasn't sure what's the right way to get it. Is there a difference between the two? Is it OK if I replace it everywhere I get the framework out of a dependency?

davidfowl replied:
That's the right way.

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
+                return true;
+            }
+
+            if (compilerIO.Outputs.Any(o => ! File.Exists(o)))
```
nit; The space after `! File.Exists()` should be removed (all over the code base)

cdmihai commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -248,6 +293,103 @@ private static void Sort(ProjectDescription project, Dictionary<string, ProjectD
+                Debug.Assert(File.Exists(input), $"Input {input} in project {project.ProjectName()} does not exist");
+            }
+
+            return compilerIO;
```
The above assertions need to be moved into NeedsRebuilding and force a rebuild.

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
+
+            foreach (var outputPath in compilerIO.Outputs)
+            {
+                if (File.GetLastWriteTime(outputPath) >= minDate) continue;
```
No self closing if statements. Just add braces

cdmihai commented:
@dotnet-bot test this please

cdmihai commented:
Addressed David's comments (except the ones that still show up in the PR; They'll go in future work once incremental story is more mature)

davidfowl commented:
@cdmihai what;s the deal with this?

cdmihai commented:
@davidfowl 
This PR introduces incremental compilation that passes the basic E2E helloworld tests. At the end of this week I'll post another PR that enables the entire CLI to build itself incrementally.

Regarding your prior feedback that I did not fix yet:

- NeedsRebuild needs some refactoring, true; However, for now, I am working on making sure CLI selfhosts incrementaly (this week's milestone) and then, during testing and stabilization I want to refactor things (fix as many of the todos in the code as possible)

Concerning CompilerUtil, it's purpose is to contain the functionality that's shared between build and compile (build needs to call it to compute incremental checks and make decisions on when to invoke compile). It consists of code I moved from inside Compile to CompilerUtil. Ideally, this would be an interface that both Compile and Build use. Since all the utility methods receive a ProjectContext, the utility methods could also compile / build specific extension methods on ProjectContext. Even more ideally, if we want to make the entire CLI incremental, there would be some higher level interface that all verbs need to implement. For now, I want to keep CompilerUtil as it is to document the shared functionality and decide later (during testing and stabilization) on how to best refactor it (maybe I shouold rename it to something else to make it clear that it only contains shared stuff).

brthor commented on scripts/test/runtests.sh:
```diff
@@ -6,6 +6,12 @@
 set -e
 
+#force tests to use stage 2 output
+export StartPath=$PATH
```
This file should be agnostic to which version of the binaries it is testing. 

It will be used to test binaries in packages and other places than the stage2 directory. 
This logic should be in the calling script.

piotrpMSFT replied:
Agreed

brthor commented on scripts/test/runtests.ps1:
```diff
@@ -5,6 +5,11 @@
 . "$PSScriptRoot\..\common\_common.ps1"
 
+#force tests to use stage 2 output
+$StartPath = $env:PATH
```
This file should be agnostic to which version of the binaries it is testing.

brthor commented on test/E2E/E2ETest.cs:
```diff
@@ -50,6 +51,78 @@ public void TestDotnetCompile()
+            TestSetup();
+
+            //first build
+            TestRunCommand("dotnet", $" --verbose build --build-profile -o {OutputDirectory}");
```
Please update these tests to use the Test infrastructure similar to the new factoring in E2E

brthor replied:
This should not be too bad, but I realized the --build-profile flag will need to be added to BuildCommand.cs. All other flags should be there:

https://github.com/dotnet/cli/blob/master/test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/BuildCommand.cs

cdmihai commented:
@brthor @davidfowl 

Regarding running tests for stage2. Should I do a different PR where I solve that issue, since the bug was there before this PR, but is blocking it?

brthor commented:
@cdmihai What is the bug exactly? 

I'm not against forcing the tests to run against stage2, I just think that the path should be set before calling the `runtests` scripts so that script becomes reusable.

For example, I am hoping to use that script to run all tests against the binaries installed by the debian package as a sort of verification against packaging bugs.

davidfowl commented:
We need to merge something as soon as possible because the ASP.NET builds are hella slowing down nexus of the CLi switch. We also need to allow passing multiple projects to build so we can build thr graph in a single call.

piotrpMSFT commented on scripts/test/runtests.sh:
```diff
@@ -42,7 +48,7 @@ failCount=0
 for project in ${TestProjects[@]}
 do
-    ./corerun  "xunit.console.netcore.exe" "$project.dll" -xml "project.xml" -notrait category=failing
+    ./corerun  "xunit.console.netcore.exe" "$project.dll" -xml "project.xml" -notrait category=failing -parallel none
```
Parallel on plz

piotrpMSFT commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -14,6 +15,7 @@
 
 namespace Microsoft.DotNet.Tools.Build
 {
+    // todo: Convert CompileContext into a DAG of dependencies: if a node needs recompilation, the entire path up to root needs compilation
```
Issue #?

cdmihai commented:
@brthor 

On the powershell scripts, stage0 is used to run the tests, not stage2. I'll fixing that (telling runntests.ps1 which stage to test) in a separate PR as suggested.

cdmihai commented:
Addressed feedback:

- tests are run in parallel
- separated passing the bin dir to test in another PR: #748
- rewrote the incremental tests to use the new testing framework

Currently blocked by #748. Can merge once that PR is in.

cdmihai commented:
@dotnet-bot test this please

cdmihai commented:
@dotnet-bot test this please

cdmihai commented:
Closing this PR. It is subsumed by #761
